Make `step.fetch` support custom retry header parsing logic

demosjarco opened the pull request:
Add support for custom retry header parsing logic in `step.fetch`.

* Update `DurableFetchConfig` interface to allow a callback function for `followRetryAfter`.
* Modify `fetch` method to check if `followRetryAfter` is a function and use it to parse the retry header.
* Add logic to handle custom retry header parsing using the provided callback function.
* Update example in `examples/fetch/src/index.ts` to demonstrate the usage of the custom retry header parsing logic.

LuisDuarte1 commented on packages/core/src/lib.ts:
```diff
@@ -499,6 +499,11 @@ export class FlowcesinhaContextBase<
 				}
 
+				if (typeof followRetryAfter === "function") {
+					const retryAfterDate = followRetryAfter(response.headers);
+					return retryAfterDate;
```
Maybe we should try-catch the `followRetryAfter` call in case it errors out (and propagate errors to the error reporter, if it happens) and return undefined like [in here](https://github.com/cloudflare/workers-sdk/blob/f9fd9df8f6e11d87bb34ed5005730de1d593989a/packages/miniflare/src/runtime/config/generated.ts#L2009), wdyt?

demosjarco replied:
What should be the intended behavior if it does error out? Try to guess the retry header? Fall back explicitly to `Retry-After`? Or just no retry logic at all?

demosjarco commented:
I added the ability for falling back to header parsing if function parsing doesn't work (or we can skip it).
I also added in the ability to pass in just a string (the header name to look for) in case it doesn't need special parsing, but just a different name (like discord).
